### ISSUE TITLE
Stepped Slider: Followup changes and improvements

### ISSUE THREAD

**Issue by jasmussen** [WordPress/gutenberg#43412]
A stepped slider recently landed, allowing you to set padding in multiples of a base unit. There are a few followups to look at.

## Fix artifact glitches 🚨

* [x] Fix artifact glitches with knob

Occasionally (exact steps to reproduce are unclear) but there's occasionally a white artifact on top of the blue knob:

<img width="317" alt="Screenshot 2022-08-19 at 10 44 47" src="https://user-images.githubusercontent.com/1204802/185592492-77789fc6-4fdd-4de6-92ee-839679adc98d.png">

## Update the design to remove the knob

* [ ] Move to a segment selection interface instead of the knob

Placing the handle at the end of the current value makes sense when we think of this as a traditional range, but it might be better for each step to act as a handle:

<img src="https://user-images.githubusercontent.com/846565/202722831-6c84a831-828c-4287-a358-598e4dc3c8a2.gif" width="280">

This makes it a bit more obvious that each step acts as an individual button for quicker jumping.

## Increase density when unlinked

- [x] https://github.com/WordPress/gutenberg/issues/49264

The unlinked splits each value into a single row which occupies a lot of space in the context of other controls. We can remedy this by splitting the layout into two columns:

<img width="292" alt="Screenshot 2023-02-17 at 10 32 47" src="https://user-images.githubusercontent.com/846565/219620582-4999e7c5-dfe2-4270-b11a-591f492fc503.png">

## Nice to have: Unit multiplier in the padding visualization

* [ ] Show the unit multiplier in the padding visualization, top/right/bottom/left on the padding silhouette

<img width="551" alt="Screenshot 2022-09-16 at 16 05 43" src="https://user-images.githubusercontent.com/1204802/190658042-440879a4-ce2f-4522-aa9a-998cf93913b5.png">

---

Issue updated Nov 21.

<details>
<summary>Completed</summary>

## Replace the labels with numbers

* [x] Replace the small/medium/large labels numbers 0-7 (#44247)

The values 0-7 replace 0, 2x-small, x-small, small, medium, large, x-large, 2x-large

<img width="444" alt="Spacing" src="https://user-images.githubusercontent.com/1204802/190656945-beeeeedc-7ffd-4537-90c0-8f6825f3cde8.png">

## Update spacing and labels

* [x] https://github.com/WordPress/gutenberg/issues/44121 🚨 (#44141)
* [x] Individual side labels should use the new all-caps label style 🚨 (also covered in #44141)

<details>

**Comment by jameskoster** [comment]
> So we have these labels: 0, 1x, 2x, 3x, 4x, 5x, 6x, 7x

This convention works well for additive scales, e.g. an additive scale with a base of 8px creates:

* 1x = 8
* 2x = 16
* 3x = 24
* 4x = 32

Which makes sense. 

However, when an increment is used to calculate values sequentially it falls down a bit. e.g: base= 8 + Increment= 2 creates:

* 1x = 8
* 2x = 16 (2x8)
* 3x = 32 (2x16)
* 4x = 64 (2x 32)

Here the 'x' doesn't make much sense. 

---

Do we need the 'x' at all? Simple '1', '2', '3', '4' labels might be just as useful as '1x', '2x', etc, and work equally well for both types of scale.

**Comment by jameskoster** [comment]
One other observation – the label in this control is using outdated type styles. Compare 'Padding' and 'Margin' to 'Size', 'Block spacing', and 'Radius':

<img width="289" alt="Screenshot 2022-08-19 at 16 16 37" src="https://user-images.githubusercontent.com/846565/185651342-aecaf06f-fadc-4802-a864-4e0c8da8f5dc.png">

**Comment by glendaviesnz** [comment]
>Don’t show a value label, but show the multiplication operation we’re doing. Don’t include the unit (rem) — just the increment. The actual multiplication.

Just a note that we would need to do some detection to work out what to show here, as it is possible for theme authors to add a static list of values that aren't calculated, in which case we would have to default back to the increment names. 

As @jameskoster notes it is also possible to use `+` increment instead of a `*` so would need to work out how to indicate that as well. I don't think `1,2,3,4,5` alone is a great option - it didn't seem to be well accepted on the font size control which has gone back to t-shirt sizes ... which maybe is a reason to stick with the t-shirt sizes here for consistency?

**Comment by jasmussen** [comment]
> Just a note that we would need to do some detection to work out what to show here, as it is possible for theme authors to add a static list of values that aren't calculated, in which case we would have to default back to the increment names.

Should this even be allowed? It seems like in these cases we should just fall back to the custom value control.

**Comment by glendaviesnz** [comment]
> Should this even be allowed? 

Yep, absolutely - there was  discussion [with theme authors about this here](https://github.com/WordPress/gutenberg/issues/39371) - actual comments may be collapsed now, but there was a clear indication from them that they wanted the flexibility to define their own custom list of presets that may or may not consist of a calculable scale (eg. some are already using there own set list in order to support fluid clamp values which are not currently supported in the UI). A custom scale fall back is not an alternative in their case as they want to limit their users to a set list of options.

**Comment by jasmussen** [comment]
Right, makes sense, but I could see that being a dropdown list like it is for font sizes. The segmented slider to me indicates a clear delineation of values, and that assigning random properties on such a linear scale is likely going to confuse. A dropdown fallback is something we can explore, but it doesn't seem urgent to optimize the stepped slider for arbitrary values.

**Comment by glendaviesnz** [comment]
Yeh, for sure, not a blocker to looking at improvements, was just pointing it out so it was not overlooked as a requirement.

**Comment by aaronrobertshaw** [comment]
Thanks for exploring these follow-ups 👍 

There's another aspect to this control that I think we could explore improving: the UI's complexity when manipulating all sides. 

It wasn't too long ago we were undertaking a real push to clean up the sidebar and reduce clutter.  Even with only the padding control in an expanded state, the dimensions panel gets very busy quickly. Once we also have margin support, that problem is at least doubled. This is before blocks begin to add further controls there such as height, width etc.

Is it worth exploring using a popover for the expanded/unlinked view?

<img width="284" alt="Screen Shot 2022-08-22 at 4 48 19 pm" src="https://user-images.githubusercontent.com/60436221/185871165-7f1cbb0f-5eda-4d34-83c6-d2a3a87888fc.png">

**Comment by jasmussen** [comment]
Yes! 100% that should be improved. It deserves some good ideas with mockups, plus its own ticket. A few ideas to consider for precedence, such as the border control: 

<img width="279" alt="Screenshot 2022-08-22 at 10 10 52" src="https://user-images.githubusercontent.com/1204802/185872287-d9fc517c-dc3e-41f8-a199-6c553fdf3d2c.png">

Or perhaps even a more basic layout a la Figma:

![figma](https://user-images.githubusercontent.com/1204802/185872599-bf777b26-3c40-48c9-bf92-1bd80b7cf9a3.gif)

This older (outdated) mockup splits it more simply, like so:

<img width="688" alt="Padding" src="https://user-images.githubusercontent.com/1204802/185872764-2061a1f3-b1d3-42a7-b4ee-d9157dd357d2.png">

In all these cases, the stepped slider by virtue of its size may need to be invoked in a popover. But could we start with a stepper?

**Comment by ndiego** [comment]
Hi all, I have a question about the actual variable names that are calculated. Perhaps I missed a previous comment, but it's a bit weird why the variables are 30, 40, 50, etc. when that does not match the actual value. See the screenshot below. Maybe something more generic, like 1, 2, 3, 4, etc? I do like how the variables are not tied to the value, as this allows you to do some cool things based on device size. 

![image](https://user-images.githubusercontent.com/4832319/186021047-41305909-fd05-49b1-963b-db45f8c31458.png)

**Comment by glendaviesnz** [comment]
> but it's a bit weird why the variables are 30, 40, 50, etc.

@ndiego there was a lot of discussion around this [here](https://github.com/WordPress/gutenberg/issues/39371) (some of the comments around it are probably hidden in that thread now as it was getting quite long). The choice of these slugs was based around wanting to make it easier for themes to insert additional options between the Core provided values, eg if a theme wants an extra value between `30` and `40` they can add `35` - which is better than `3.5`. The choice of numbers here was also to allow the ability to easily fallback to `nearest` option if block content/patterns are copied between themes that don't support the same list of presets, eg. `--wp--preset--spacing-35` can be mapped to  `--wp--preset--spacing-30`.

**Comment by glendaviesnz** [comment]
> eg if a theme wants an extra value between 30 and 40 they can add 35 

Except I have just noticed that values are merged into the Core settings for the generation of the CSS properties, but not at the UI level for the selection of the values, so need to take a look at that. `fontSizes` does the same, merges the theme and core values to produce the presets CSS properties, but then only shows the theme values to select from in the UI - not sure what the thinking is behind this 🤔

**Comment by jasmussen** [comment]
Extracting from a conversation [here](https://github.com/WordPress/gutenberg/issues/44133#issuecomment-1246416237), let’s move forward with the numbers as spacing indicators rather than "1x", or the current "small/medium/large" etc. This is mainly important for any CSS classes we mean to extrapolate from the value, which should be along the lines of `has-margin-s1` or `has-spacing-1`, rather than `has-spacing-small`. Here's how that would look:

<img width="444" alt="Spacing" src="https://user-images.githubusercontent.com/1204802/190656945-beeeeedc-7ffd-4537-90c0-8f6825f3cde8.png">

**Comment by glendaviesnz** [comment]
> Extracting from a conversation [here](https://github.com/WordPress/gutenberg/issues/44133#issuecomment-1246416237), let’s move forward with the numbers as spacing indicators rather than "1x", or the current "small/medium/large" etc. 

@jasmussen there is [a PR that implements this here](https://github.com/WordPress/gutenberg/pull/44247).


>This is mainly important for any CSS classes we mean to extrapolate from the value, which should be along the lines of `has-margin-s1` or `has-spacing-1`, rather than `has-spacing-small`.

Just to clarify, there are currently no CSS classes as part of the first release of the spacing presets in 6.1, they are all applied via inline styles and CSS vars, and these are extrapolated from the `slug` value, not the name/label that appears in the UI. [As noted here](https://github.com/WordPress/gutenberg/issues/43412#issuecomment-1223334282) the `slugs` are of the format `10,20,30,40`. 

The spacingSizes are generated up and down from a midpoint of `50` in order to allow for better cross-site/theme fallbacks if copied blocks or patterns use spacing preset values not supported by the current theme, ie. if different spacing scales all have a common `medium` point in the scale we are more likely to get a comparable match when content is moved between sites. The one drawback of switching from t-shirt sizing to digits is that there now appears to be a vague correlation between the label and the slug, eg. on this PR if you set a value of `1` you will get a CSS var of `var(--wp--preset--spacing--20)` applied as the default Core scale only has 3 steps either side of the `50` medium step. This may or may not be a problem, but could potentially be slightly more confusing than a `2x-small` label and `20` slug combination where there is no correlation between the label and slug.

We can't add CSS classes until we have the infrastructure in place to apply them dynamically on the front end based on the current blocks as there need to be a huge number of them, eg. `has-spacing` doesn't work as there need to be classes for each spacing type, eg. `has-margin`, `has-padding`, etc. and for each side, `has-margin-left, has-margin-right` and for each spacing size, `has-margin-left-10, has-margin-left-20`, etc.

**Comment by glendaviesnz** [comment]
I think everything that was planned for 6.1 is done now, so removing from the 6.1 editor board

**Comment by jameskoster** [comment]
Placing the handle at the end of the current value makes sense when we think of this as a traditional range, but it might be better for each step to act as a handle:

<img src="https://user-images.githubusercontent.com/846565/202722831-6c84a831-828c-4287-a358-598e4dc3c8a2.gif" width="280">

This makes it a bit more obvious that each step acts as an individual button for quicker jumping.

**Comment by jasmussen** [comment]
I updated the issue to include that mockup, Jay!

**Comment by jasmussen** [comment]
I've conditionally checked the box for the visual glitch when dragging in Safari, mainly because I can no longer reproduce it in trunk. _If you can, if you still see that white glitch, please do uncheck the box again and ideally share reproduction notesd, thank you!_

**Comment by annezazu** [comment]
👋🏼 I'm doing a [sweep of high priority labeled issues](https://wordpress.slack.com/archives/C02QB2JS7/p1702996696603749) and noticed this has greatly stalled. This label is meant for high priority items that require quick attention to resolve. To keep this label actionable and high impact, I'm culling some of the older, stalled items so we can revisit. If you believe the label should be added back, feel free to do so/just let me know and let's do what we can to mobilize efforts. As it stands though, this [isn't on the 6.5 roadmap ](https://make.wordpress.org/core/2023/12/07/roadmap-to-6-5/)and doesn't reflect current high priority items from what I can see across the project.

**Comment by jasmussen** [comment]
@aduth Based on latest component work, what's your instinct for what to do with the remaining two issues on this one? Thanks for looking.

**Comment by aduth** [comment]
I'd be curious for @mirka 's thoughts on whether there's a plan to include this kind of step behavior as part of the `Slider` component planned in #74178 . If so, it might be better to consider these improvements as part of that work. 

The unit value being shown in the silhouette preview feels like more of a separate editor question than a component one. Personally I don't have a strong opinion one way or the other.

If the question is "can we close this", then yeah, my sense is that it'd make more sense to prioritize work on form components in `@wordpress/ui` and consider any improvements as part of that. This would be covered by #74178.

**Comment by jasmussen** [comment]
Cool, I'm going to close this one for now, there's plenty of opportunity for others to reopen if this was in error.

**Comment by mirka** [comment]
> whether there's a plan to include this kind of step behavior as part of the `Slider` component planned in [#74178](https://github.com/WordPress/gutenberg/issues/74178)

Yes!
